File: client/src/models/Transfer.ts
import type Admin from "./Admin";

type TransferJSON = {
  id?: string;
  type?: string;
  amount?: number;
  agent_id?: string;
  admin_id?: string;
  created_at?: string;
};
class Transfer {
  #id: string;
  type: "fund" | "refund";
  #amount: number;
  #agent_id: string;
  #admin_id: string;
  #admin_ref: Admin | null;
  created_at: string;
  constructor(t: TransferJSON, admin?: Admin) {
    this.#id = t.id || "";
    this.type = t.type === "fund" ? "fund" : "refund";
    this.#amount = t.amount || 0;
    this.#admin_ref = admin || null;
    this.#agent_id = t.agent_id || "";
    this.#admin_id = t.admin_id || "";
    this.created_at = t.created_at || "";
  }
  get id(): string {
    return this.#id;
  }
  get amount(): number {
    return this.#amount;
  }
  get agent_id(): string {
    return this.#agent_id;
  }
  get admin_id(): string {
    return this.#admin_id;
  }
  get admin(): Admin | null {
    return this.#admin_ref;
  }
  get date(): Date {
    return new Date(this.created_at);
  }
}
export { type TransferJSON };
export default Transfer;

File: server/models/transfer.go
package models

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type Transfer struct {
	ID        uuid.UUID `gorm:"column:id;type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	Type      Status    `gorm:"column:type;not null;type:varchar(20)" json:"type"` // Deposit / Withdraw
	Amount    uint      `gorm:"column:amount;not null" json:"amount"`
	AdminID   uuid.UUID `gorm:"column:admin_id;type:uuid;index" json:"admin_id"`
	AgentID   uuid.UUID `gorm:"column:agent_id;type:uuid;index" json:"agent_id"`
	CreatedAt time.Time `gorm:"column:created_at;autoCreateTime" json:"created_at"`

	Admin *Admin `gorm:"foreignKey:AdminID" json:"admin"`
	Agent *Agent `gorm:"foreignKey:AgentID" json:"agent"`
}

func (Transfer) TableName() string {
	return "transfer"
}

func (t Transfer) String() string {

	str := fmt.Sprintf("Transfer(id=%s, type='%s', amount=%d, admin_id='%s', agent='%s'", t.ID, t.Type, t.Amount, t.AdminID, t.Agent)
	str += ")"
	return str
}
